subsys/net/lib/tftp/tftp_client.c

```diff
@@ -0,0 +1,524 @@
+		 * -> Before pushing data to the buffer, we should check if we
+		 * have overflowed? If so, we can probably store some of this
+		 * data and exit after sending an error to the server. Or we
+		 * can simply exit without storing any data. Both approaches are
+		 * okay. Here, we've implemented the first one.
+		 */
+		if ((tftpc_request_size - TFTP_HEADER_SIZE) >
+				(client->user_buf_size - tftpc_index)) {
+			/* Can't copy all. Lets only copy a minimal set and
+			 * set an error.
+			 */
+			cpy_size = client->user_buf_size - tftpc_index;
+			sendack = false;
+		} else {
+			/* Can copy all. */
+			cpy_size = tftpc_request_size - TFTP_HEADER_SIZE;
+		}
+
+		/* Perform the actual copy and update the index. */
+		memcpy(client->user_buf + tftpc_index,
+			tftpc_request_buffer + TFTP_HEADER_SIZE, cpy_size);
+		tftpc_index += cpy_size;
+
+		/* Was there any problem? */
+		if (sendack == true) {
+
+			/* Now we are in a position to ack this data. */
+			send_ack(sock, block_no);
+		} else {
+
+			/* Send Error :-( */
+			send_err(sock, 0x3, "Buffer Overflow");
+
+			/* Log the error. */
+			LOG_ERR("For the transfer of this file, the user provided \
+				  buffer of size: %d", client->user_buf_size);
+			LOG_ERR("However, it seems that the actual file has larger \
+				  size. Exiting !");
+
+			return TFTPC_BUFFER_OVERFLOW;
+		}
+
+		/* We recevied a "block" of data. Lets update the global
+		 * book-keeping variable that tracks the number of blocks
+		 * received.
+		 */
+		tftpc_block_no++;
+
+		/* Because we are RFC1350 compliant, the block size will
+		 * always be 512. If it is equal to 512, we will assume
+		 * that the transfer is still in progress. If we have
+		 * block size less than 512, we will conclude the transfer.
+		 */
+		return (tftpc_request_size - TFTP_HEADER_SIZE ==
+				TFTP_BLOCK_SIZE) ?
+				(TFTPC_DATA_RX_SUCCESS) :
+				(TFTPC_OP_COMPLETED);
+	} else if (tftpc_block_no > block_no) {
+
+		LOG_INF("Server send duplicate (already acked) block again. \
+				 Block Number: %d", block_no);
+		LOG_INF("Client already trying to receive Block Number: \
+				 %d", tftpc_block_no);
+
+		/* Re-ACK the data. */
+		send_ack(sock, block_no);
+
+		/* OK - This means that we just received a block of data
+		 * that has already been received and acked by the TFTP
+		 * client.
+		 */
+		return TFTPC_DUPLICATE_DATA;
+	}
+
+	/* Don't know what's going on. */
+	return TFTPC_UNKNOWN_FAILURE;
+}
+
+/* Name: tftp_send_request
+ * Description: This function sends out a request to the TFTP Server
+ * (Read / Write) and waits for a response. Once we get some response
+ * from the server, it is interpreted and ensured to be correct.
+ * If not, we keep on poking the server for data until we eventually
+ * give up.
+ */
+static int tftp_send_request(int sock, u8_t request,
+		const char *remote_file, const char *mode)
+{
+	u8_t    no_of_retransmists = 0;
+	s32_t   stat;
+	u16_t   server_response    = -1;
```

Comment:
Please arrange the variables so that the assignment vars are first, and then the rest. So stat var after server_response.